### PR TITLE
fix: normalize MCP tool schemas for Gemini

### LINKED ISSUE #11162: Google AI Studio (Gemini) 400 Error: "property is not defined" due to invalid MCP tool schema (e.g. serial.open)
### Plugin Type

VSCode Extension

### Cline Version

3.86.0

### Beta version

- [ ] I am using a beta version of Cline

### What happened?

When using Gemini 3.1 Pro Preview via the Google AI Studio provider, the model returns a 400 Invalid Argument error. This happens because Google AI Studio strictly validates the JSON schemas of available tools. In this case, one of the registered MCP tools (`serial.open` from `serial-mcp-server`) has an invalid schema. Specifically, it lists `"port"` in the `required` array, but the `properties` object is completely missing from the schema definition. 

Expected behavior: Cline should either gracefully handle invalid JSON schemas from MCP servers (perhaps sanitizing or automatically patching missing empty `properties` objects) so that strict providers like Google AI Studio do not crash the entire request, or surface a clear error targeting the specific MCP server that is outputting an invalid schema.

### Steps to reproduce

1. Start Cline and configure the provider to `Google AI Studio` and select a Gemini model.
2. Ensure you have an MCP server connected that provides an invalid JSON schema. Specifically, the `serial-mcp-server` offering the `serial.open` tool which has `{ "required": ["port"], "type": "OBJECT" }` but no `"properties"` definition.
3. Attempt any prompt or task that causes the tool definitions to be sent to the Gemini API.
4. The request will fail with a 400 error stating that a required property is not defined.

### Provider/Model

openai / google/gemini-3.1-pro-preview

### System Information

Visual Studio Code: 1.122.1, Node.js: v22.22.1, Architecture: x64

### PR DESCRIPTION
**Pull request by he-yufeng**
﻿## Summary
- normalize AI SDK tool schemas before they are sent to Gemini
- lower-case object schema types from MCP servers that emit `OBJECT`
- define placeholder property schemas for required fields that are missing from `properties`

Fixes #11162.

## To verify
- `bun -F @cline/llms test src/providers/gateway.test.ts -t "normalizes MCP object schemas"`
- `bun -F @cline/llms typecheck`
- `bun -F @cline/llms build`
- `bun biome check --diagnostic-level=error packages/llms/src/providers/ai-sdk.ts packages/llms/src/providers/gateway.test.ts`

---

> [!NOTE]
> **Low Risk**
> Localized change to tool schema shaping on the AI SDK path; behavior is covered by a new Gemini normalization test and does not touch auth or persistence.
> 
> **Overview**
> **Tool schemas from MCP servers are normalized** before they are passed to the AI SDK (`jsonSchema` in `toAiSdkTools`), so Gemini and similar providers receive valid JSON Schema instead of MCP-specific shapes.
> 
> The updated **`normalizeAiSdkToolInputSchema`** lowercases string `type` values (e.g. `OBJECT` → `object`), recursively normalizes nested property schemas, and treats schemas with `properties` / `required` / `additionalProperties` as object-like even when `type` is wrong. **Required field names** that are missing from `properties` get a placeholder `{}` entry (e.g. `port` when only nested fields were listed). Invalid `required: true` on a property node is no longer copied into the output schema.
> 
> A **gateway test** asserts Gemini receives the normalized schema for a representative MCP tool definition.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cb6b804f9672d5b9f3f9f42fee4d615136543318. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by he-yufeng** [comment]
Updated this PR to cover the two schema-normalization edge cases from the automated review: nested property schemas are normalized recursively, and non-array `required` values are dropped instead of being forwarded to the provider.

Re-ran:
- `bun -F @cline/llms test -- src/providers/gateway.test.ts -t "normalizes MCP object schemas before sending them to Gemini"` — 1 passed
- `bun -F @cline/llms typecheck` — passed
- `bunx --bun @biomejs/biome check --diagnostic-level=error sdk\packages\llms\src\providers\ai-sdk.ts sdk\packages\llms\src\providers\gateway.test.ts` — passed
- `git diff --check -- sdk\packages\llms\src\providers\ai-sdk.ts sdk\packages\llms\src\providers\gateway.test.ts` — passed

**Review by dominiccooney — CHANGES_REQUESTED** [review]
The reported top-level case is fixed, but recursive normalization stops at `properties`.

`normalizeAiSdkToolInputSchema()` returns as soon as the current node is not object-like. For example, after it lowercases an array property from `ARRAY` to `array`, it returns that node without visiting `items`; `anyOf`, `oneOf`, and `allOf` branches are likewise never visited. A malformed MCP schema such as `{ type: "OBJECT", properties: { rows: { type: "ARRAY", items: { type: "OBJECT", properties: { id: { type: "STRING" } } } } } }` still sends uppercase nested types to Gemini and can fail the whole model request.

Please make the compatibility transform recurse through the schema-valued locations in the declared JSON Schema contract—not arbitrary object values, which may be examples/defaults containing a data field named `type`. At minimum cover object properties/pattern properties/additional-properties schemas, array items/prefix items, composition branches (`anyOf`/`oneOf`/`allOf`), negation/condition branches, and definitions. Add a provider-boundary test with an array and a composition, plus a canonical-schema case showing valid input is unchanged. The existing reported-case test should remain.

I attempted the focused test on a current-main merge, but this isolated review worktree lacked the SDK packages' per-package dependency links and failed before collecting tests (`zod` could not be resolved from `@cline/shared`). That is a review-environment setup issue, not a failure of this branch.

File: sdk/packages/llms/src/providers/ai-sdk.ts [hunk 1]
```diff
@@ -452,13 +456,46 @@ export async function repairMalformedToolCall<T extends RepairableToolCall>({
...
+		return schema;
+	}
+
+	const properties: Record<string, unknown> = {};
+	if (isRecord(schema.properties)) {
+		for (const [key, value] of Object.entries(schema.properties)) {
+			properties[key] = isRecord(value)
+				? normalizeAiSdkToolInputSchema(value)
+				: value;
+		}
```
[the remark this answers is not in the document]
Comment (he-yufeng) [reply sdk/packages/llms/src/providers/ai-sdk.ts:504]: Fixed in 2c1a7f7d7: normalization now recurses into `items` (object and tuple forms) and into `anyOf`/`oneOf`/`allOf` branches before the object-likeness gate, so nested uppercase types lowercase at every depth. Covered by the new "normalizes nested array items and combinator branches" test.

File: sdk/packages/llms/src/providers/ai-sdk.ts [hunk 2]
```diff
@@ -452,13 +456,46 @@ export async function repairMalformedToolCall<T extends RepairableToolCall>({
...
+
+	const isObjectLike =
+		schema.type === "object" ||
+		"properties" in schema ||
+		"required" in schema ||
+		"additionalProperties" in schema;
+
+	if (!isObjectLike) {
+		return schema;
+	}
```
[the remark this answers is not in the document]
Comment (he-yufeng) [reply sdk/packages/llms/src/providers/ai-sdk.ts:496]: Fixed in 595b17fec: the object-likeness early return now applies only below the root. A bare or scalar-only root schema (empty MCP inputSchema included) still comes out as `{ type: "object" }`, matching what Gemini requires for tool parameters. Covered by the new root-default test.

**Comment by he-yufeng** [comment]
@dominiccooney Your recursive case is fixed in 2c1a7f7d7: normalization now recurses into items (both single-schema and tuple forms) and the anyOf/oneOf/allOf branches before the object-likeness gate, and your exact rows/items shape is the new regression test. While in there I also closed the typeless-root gap cursor[bot] flagged: a bare {} inputSchema still becomes { type: "object" } at the root (595b17fec). 92 tests green on the llms package, typecheck clean.
